lib/internal/process/parse_args.js

```diff
@@ -0,0 +1,112 @@
+'use strict';
+
+const { Array, Set, Number } = primordials;
+const { ERR_INVALID_ARG_TYPE } = require('internal/errors').codes;
+
+/**
+ * Returns an Object representation of command-line arguments to a script.
+ *
+ * Default behavior:
+ * - All arguments are considered "boolean flags"; in the returned
+ *   object, the key is the argument (if present), and the value will be `true`.
+ * - Uses `process.argv.slice(2)`, but can accept an explicit array of strings.
+ * - Argument(s) specified in `opts.expectsValue` will have `string` values
+ *   instead of `true`; the subsequent item in the `argv` array will be consumed
+ *   if a `=` is not used.
+ * - If `=` is present, the value will be boolean if and only if the RHS of the
+ *   string is one of string `true` or `false`
+ * - "Bare" arguments are those which do not begin with a `-` or `--` and those
+ *   after a bare `--`; these will be returned as items of the `_` array
+ * - The `_` array will always be present, even if empty.
+ * - Repeated arguments are combined into an array
+ * @param {string[]} [argv=process.argv.slice(2)] - Array of script arguments as
+ * strings
+ * @param {Object} [options] - Options
+ * @param {string[]|string} [opts.expectsValue] - Argument(s) listed here expect
+ * a value
+ * @example
+ * parseArgs(['--foo', '--bar']) // {foo: true, bar: true, _: []}
+ * parseArgs(['--foo', '-b']) // {foo: true, b: true, _: []}
+ * parseArgs(['---foo']) // {foo: true, _: []}
+ * parseArgs(['--foo=bar']) // {foo: true, _: []}
+ * parseArgs(['--foo', 'bar']) // {foo: true, _: ['bar']}
+ * parseArgs(['--foo', 'bar'], {expectsValue: 'foo'}) // {foo: 'bar', _: []}
+ * parseArgs(['foo']) // {_: ['foo']}
+ * parseArgs(['--foo', '--', '--bar']) // {foo: true, _: ['--bar']}
+ * parseArgs(['--foo=bar', '--foo=baz']) // {foo: ['bar', 'baz'], _: []}
+ * parseArgs(['--foo', '--foo']) // {foo: true, _: []}
+ * pasreArgs(['--foo=true']) // {foo: true, _: []}
+ * parseArgs(['--foo=false']) // {foo: false, _: [])}
+ * parseArgs(['--foo=true', '--foo=false']) // {foo: false, _: [])}
+ * parseArgs(['--foo', 'true']) // {foo: true, _: ['true']}
+ * parseArgs(['--foo', 'true'], {expectsValue: ['foo']}) // {foo: true, _: []}
+ */
+const parseArgs = (
+  argv = process.argv.slice(2),
+  options = { expectsValue: [] }
+) => {
+  if (!Array.isArray(argv)) {
+    options = argv;
+    argv = process.argv.slice(2);
+  }
+  if (typeof options !== 'object' || options === null) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'options',
+      'object',
+      options);
+  }
+  if (typeof options.expectsValue === 'string') {
```

Comment:
is there a reason for this overload? it seems somewhat nonsensical.

Reply:
It's friendly to support a string if the value can be singular, instead of requiring an array.  This sort of behavior is very common.

Reply:
that just seems confusing to me. can we remove it?

Reply:
I'd rather not!

Reply:
This overload seems to be a category error, it should be removed.

Reply:
Sorry.  This is how many, _many_ userland APIs work, is ergonomic, and trivial to support.

Reply:
I'm -0 on having this overload: it doesn't really bother me, but since we should rename `expectsValue` to something else, and if we rename to something on plural (like `optionsWithValues`) it might make more sense/be more intuitive to have this as an Array.

Reply:
For yargs' chaining API, I like that I can provide options as either an array or string, if I only have a couple options, I'll usually write:

```js
yargs
  .boolean('option-1')
  .boolean('option-2)
```

But, if I have many options that I'm configuring, I might write:

```js
yargs.boolean(['option-1', 'option-2', 'option-3', 'option-4')
```

I don't have as strong of an opinion in this case, given that this API isn't chaining like yargs (_you configure all the options in one go any ways_).

---

I don't think either implementation would be a major usability issue, and am supportive of whatever we land on.

Reply:
Haven't decided yet where I land on "accept string or not", but if we don't accept string it should throw when string is passed, otherwise we'll get situations where the string is treated as an array (`"foo"` getting treated as `['f', 'o', 'o']` leads to hard to debug issues).

Reply:
I'm -0 too. Either is fine by me, but I guess if it was up to me, I'd prefer to not permit a string initially, because adding it later is no problem but taking it away once it's out there is a big problem. That said, I doubt we'd have to/want to take it away, hence my -0 rather than -1.

Reply:
I guess I am also a member of the -0 club.